File: 0226-invert-binary-tree/0226-invert-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function invertTree(root: TreeNode | null): TreeNode | null {
   function invert(root: TreeNode | null){
        if(!root) return;
        [root.left, root.right] = [root.right, root.left];
        invert(root.left);
        invert(root.right);
    }
    invert(root);
    return root;
};

File: 0226-invert-binary-tree/0226-invert-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
type TreeRef = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn invert_tree(root: TreeRef) -> TreeRef {
        
        //helper function
        fn invert(root: & TreeRef) {
            //check that node is not null
            if let Some(node) = root {
                //borrow mutable reference
                let mut node = node.borrow_mut();
                //hold left node in a variable
                let temp = node.left.take();
                // assign swap left for right
                node.left = node.right.take();
                //swap right for left
                node.right = temp;
                //use recursion to perform the same operation for left and right nodes
                invert(&node.left);
                invert(&node.right);
            }
        }
        
        //invert the tree
        invert(&root);
        
        //return the tree
        root
    }
}

//traverse the tree, on each node, borrow mutably and swap the children
// nodes, then recursively perform the same operation on each child node